### PR TITLE
Implement London EVMVersion and the BASEFEE opcode

### LINKED ISSUE #11390: Implement support for London
Most of the [London](https://github.com/ethereum/eth1.0-specs/blob/master/network-upgrades/mainnet-upgrades/london.md) changes will not affect us (based on the [latest testnet](https://github.com/ethereum/eth1.0-specs/blob/master/network-upgrades/client-integration-testnets/baikal.md):
- [EIP-1559](https://eips.ethereum.org/EIPS/eip-1559)
- [EIP-3198](https://eips.ethereum.org/EIPS/eip-3198)
- [EIP-3529](https://eips.ethereum.org/EIPS/eip-3529)
- [EIP-3541](https://eips.ethereum.org/EIPS/eip-3541)

1559 changes gas price / block gas limit calculation, don't think we need to do anything with this.

3198  introduces a new `BASEFEE` opcode, we may want to expose this as `block.basefee`

3529 reduces gas refunds on `selfdestruct` and `sstore`, we may want to update our documentation on this regard.

3541 runtime code starting with the `0xef` byte can not be deployed, we could display a warning for this hypothetical case, but it won't be generated by our assembler anyway.

### PR DESCRIPTION
**Pull request by hrkrshnn**
Fixes #11390

Need to fix evmone CI issues.

### CODE REVIEW

File: test/libsolidity/semanticTests/state/block_basefee.sol
Comment (hrkrshnn) [inline test/libsolidity/semanticTests/state/block_basefee.sol]: TODO: need to fix evmone.

**Comment by axic** [comment]
I really fix that just like with Berlin, this shouldn't be a single large PR, but rather:
1) evmc/evmone
2) london support + basefee

I would merge 1) now, but would not merge 2) before the hardfork goes live (~3 weeks from now).

**Comment by hrkrshnn** [comment]
@axic I'll split this up, once I can get the CI working.

About the second part, why not add `london` to `evm-version`, while keep berlin as the default, and add `basefee`? We can change the default to london once the hardfork is live.

File: docs/cheatsheet.rst
Comment (chriseth) [inline docs/cheatsheet.rst]: Maybe we should explain basefee (at least mention it) in the documentation and link it here?

**Comment by axic** [comment]
> About the second part, why not add london to evm-version, while keep berlin as the default, and add basefee? We can change the default to london once the hardfork is live.

It worked before when we could mark those as "experimental", but you have removed that option 😅

File: test/EVMHost.cpp
Comment (axic) [inline test/EVMHost.cpp]:
```suggestion
		m_evmRevision = EVMC_LONDON;
```

How is this PR supposed to work? :sweat_smile:

File: test/evmc/evmc.h
Comment (axic) [inline test/evmc/evmc.h]: Are you sure you updated every relevant file in this directory?

Comment (hrkrshnn) [reply test/evmc/evmc.h]: I actually did that in the second PR: https://github.com/ethereum/solidity/pull/11649/files

Will rebase this, once that's done.

File: test/EVMHost.cpp
Comment (axic) [inline test/EVMHost.cpp:138]: Ok, so we missed this in #11706.

Comment (axic) [reply test/EVMHost.cpp:138]: Though it also makes logically sense in this PR, so fine either way.

**Comment by axic** [comment]
This looks good to me, though:
1) Would still wait the extra week until London goes live and only merge this after.
2) Clean up the commits a tiny bit (e.g. the commit "Updated evmone version to 0.8.0 and evmc version to 9.0.0" has nothing to do with what it says)

File: test/libyul/objectCompiler/dataoffset_data.yul
Comment (chriseth) [inline test/libyul/objectCompiler/dataoffset_data.yul:16]: :)

**Comment by chriseth** [comment]
Why wait? As long as we do not change the default evm version it should be fine.

**Comment by axic** [comment]
> Why wait? As long as we do not change the default evm version it should be fine.

Because if the fork falls through there's just extra stuff to be removed. It is really one week away so makes little difference in terms of "stalled PRs" waiting that out :)

**Comment by azf20** [comment]
Excited about this - is there any way to access the block BASEFEE before this is merged & a new version is released?

**Comment by hrkrshnn** [comment]
@azf20 Currently the only way is to use `verbatim` in Yul. There is an example in https://blog.soliditylang.org/2021/06/10/solidity-0.8.5-release-announcement/ about `basefee` using `verbatim`.

**Comment by chriseth** [comment]
... but you unfortunately cannot use it in inline assembly.

File: liblangutil/EVMVersion.h
Comment (hrkrshnn) [inline liblangutil/EVMVersion.h:104]: I'll make a separate PR changing the default to London.

File: test/EVMHost.cpp
Comment (leonardoalt) [reply test/EVMHost.cpp:138]: Why 7?

Comment (leonardoalt) [reply test/EVMHost.cpp:138]: Ah that's the cost, nevermind

File: Changelog.md
Comment (leonardoalt) [inline Changelog.md]: Maybe also add the added Yul function?

**Review by leonardoalt — COMMENTED** [review]
Minor comment, looks good

File: Changelog.md
Comment (hrkrshnn) [reply Changelog.md]: Added this below.

**Review by leonardoalt — DISMISSED** [review]

File: test/EVMHost.cpp
Comment (hrkrshnn) [reply test/EVMHost.cpp:138]: For future reference: the minimum of 7 wei is not specified in the EIP. Here's some explanation on why it's 7 wei from ETH R&D discord: https://discord.com/channels/595666850260713488/749160974506000444/861666788495851561

File: Changelog.md
Comment (chriseth) [reply Changelog.md]: Isn't this a language feature instead of a compiler feature? Both of them?

File: docs/cheatsheet.rst
Comment (chriseth) [inline docs/cheatsheet.rst]: link to the EIP?

Comment (hrkrshnn) [reply docs/cheatsheet.rst]: Added links and fixed the changelog.

File: Changelog.md
Comment (chriseth) [inline Changelog.md]:
```suggestion
 * Introduced global ``block.basefee`` for retrieving the base fee of the current block.
```
(also the code generator is affected)

File: liblangutil/EVMVersion.h
Comment (chriseth) [inline liblangutil/EVMVersion.h:94]: Needs change

Comment (hrkrshnn) [reply liblangutil/EVMVersion.h:94]: It's done in https://github.com/ethereum/solidity/pull/11763/commits/0c8b6124e3a4cb0d42e9647f5ccddfd5e5474926

File: Changelog.md
Comment (hrkrshnn) [reply Changelog.md]: Fixed this.

File: liblangutil/EVMVersion.h
Comment (hrkrshnn) [reply liblangutil/EVMVersion.h:94]: Or second commit in https://github.com/ethereum/solidity/pull/11763

File: docs/yul.rst
Comment (axic) [inline docs/yul.rst:884]: The doc on the top also explains what these letters stand for, `L` is missing.

Comment (hrkrshnn) [reply docs/yul.rst:884]: Thanks. Fixed.

File: Changelog.md
Comment (ekpyron) [inline Changelog.md]: Past tense is IMHO weird style for a changelog.

File: test/libsolidity/syntaxTests/types/magic_block_basefee_error.sol
Comment (ekpyron) [inline test/libsolidity/syntaxTests/types/magic_block_basefee_error.sol:12]: Why not
```suggestion
// EVMVersion: <london
```
or
```suggestion
// EVMVersion: <=berlin
```

**Review by ekpyron — DISMISSED** [review]
Any reason for not merging https://github.com/ethereum/solidity/pull/11763/ into this first :-)? Strictly speaking, that'd make more sense. Not that it matters too much, if we merge all of them anyways, though.

Anyways, looks fine despite minor comments.

File: test/libsolidity/syntaxTests/types/magic_block_basefee_error.sol
Comment (hrkrshnn) [inline test/libsolidity/syntaxTests/types/magic_block_basefee_error.sol:15]: @ekpyron If we change it to `<=berlin`, then the CI would fail because of differences in error messages. Berlin replaced by "Istanbul", for example.

Comment (ekpyron) [reply test/libsolidity/syntaxTests/types/magic_block_basefee_error.sol:15]: We could also make it consistent with e.g. ``"\"chainid\" is not supported by the VM version."``, which doesn't have that problem...

Comment (ekpyron) [reply test/libsolidity/syntaxTests/types/magic_block_basefee_error.sol:15]: Ah, no, that's the other one. Well, fine in any case.

**Review by ekpyron — APPROVED** [review]

File: test/libsolidity/syntaxTests/types/magic_block_basefee_error.sol
Comment (ekpyron) [reply test/libsolidity/syntaxTests/types/magic_block_basefee_error.sol:15]: Would be another reason for pulling in https://github.com/ethereum/solidity/pull/11763 first, though ;-).

**Pull request merged**
